governator-core/src/main/java/com/netflix/governator/internal/PreDestroyLifecycleActions.java

```diff
@@ -1,56 +1,146 @@
+import java.util.ArrayList;
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 
 import javax.annotation.PreDestroy;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Supplier;
 import com.netflix.governator.LifecycleAction;
+import com.netflix.governator.LifecycleFeature;
+import com.netflix.governator.internal.TypeInspector.TypeVisitor;
 
 /**
- * Special AbstractLifecycleFeature to support @PreDestroy annotation processing.
- * Note that this feature is implicit in LifecycleModule and therefore does not need 
- * to be added using the LifecycleFeature multibinding.
+ * Special LifecycleFeature to support @PreDestroy annotation processing and
+ * java.lang.AutoCloseable detection. Note that this feature is implicit in
+ * LifecycleModule and therefore does not need to be added using the
+ * LifecycleFeature multibinding.
  * 
  * @author elandau
  */
-public final class PreDestroyLifecycleActions extends AbstractLifecycleFeature {
-
+public final class PreDestroyLifecycleActions implements LifecycleFeature {
+    private static final Logger LOG = LoggerFactory.getLogger(PreDestroyLifecycleActions.class);
     public static PreDestroyLifecycleActions INSTANCE = new PreDestroyLifecycleActions();
-    
+
     private PreDestroyLifecycleActions() {
     }
 
     @Override
-    protected List<LifecycleAction> getMethodActions(final Class<?> type, final Method method) {
-        if (null != method.getAnnotation(PreDestroy.class)) {
-            method.setAccessible(true);
-            return Collections.<LifecycleAction>singletonList(new LifecycleAction() {
-                @Override
-                public void call(Object obj) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
-                    method.invoke(obj);
-                }
-                
-                @Override
-                public String toString() {
-                    return new StringBuilder()
-                        .append("PreDestroy[")
-                        .append(type.getName())
-                        .append("#")
-                        .append(method.getName())
-                        .append("]")
-                        .toString();
-                }
-            });
-        }
-        return Collections.emptyList();
+    public List<LifecycleAction> getActionsForType(final Class<?> type) {
+        return TypeInspector.accept(type, new PreDestroyVisitor());
     }
 
     @Override
     public String toString() {
         return "PreDestroy";
     }
 
+    private static class PreDestroyVisitor implements TypeVisitor, Supplier<List<LifecycleAction>> {
+        private Set<String> visitContext = new HashSet<>();
+        private List<LifecycleAction> typeActions = new ArrayList<>();
+
+        @Override
+        public boolean visit(final Class<?> clazz) {
+            boolean continueVisit = !clazz.isInterface();
```

Comment:
nit: not of fan of tracking this boolean.  I'd rather see something like 

if (!clazz.isInterface())
   return false;